atlasaction/action_test.go

```diff
@@ -87,6 +93,234 @@ func TestMigrateApply(t *testing.T) {
+	t.Run("config-broken", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("config", "file://testdata/config/broken.hcl")
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `"testdata/config/broken.hcl" was not found`)
+	})
+	t.Run("env-broken", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("config", "file://testdata/config/atlas.hcl")
+		tt.setInput("env", "broken-env")
+		tt.setInput("dir-name", "test-dir")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `env "broken-env" not defined in project file`)
+	})
+	t.Run("config", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("env", "test")
+		tt.setInput("dir-name", "test-dir")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+	})
+	t.Run("broken dir", func(t *testing.T) {
+		tt := newT(t)
+		tt.setInput("dir", "file://some_broken_dir")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `sql/migrate: stat some_broken_dir: no such file or directory`)
+	})
+	t.Run("dir-name invalid characters", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-#dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, "slug must be lowercase alphanumeric")
+	})
+	t.Run("dev-url broken", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "broken-driver://")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `unknown driver "broken-driver"`)
+	})
+	t.Run("invalid tag", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.setInput("tag", "invalid-character@")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.ErrorContains(t, err, `tag must be lowercase alphanumeric`)
+	})
+	t.Run("tag", func(t *testing.T) {
+		tt := newHttpT(t, nil)
+		tt.setupConfigWithLogin(t)
+		tt.setInput("dir", "file://testdata/migrations")
+		tt.setInput("dir-name", "test-dir")
+		tt.setInput("dev-url", "sqlite://file?mode=memory")
+		tt.setInput("tag", "valid-tag-123")
+		err := MigratePush(context.Background(), tt.cli, tt.act)
+		require.NoError(t, err)
+	})
+	t.Run("e2e test", func(t *testing.T) {
+		var syncDirCalls, pushDirCalls atomicInt
```

Comment:
Since they are scoped to this test, I don't think you need atomic ints. Did test with `-race` fail for you?

Reply:
Didn't fail, this was my first attempt since the HTTP handler runs in a different go-routine, so it made sense

I will remove and test it

Reply:
Indeed not needed